File: Algorithm/inflearn_cpp/myCode/84_휴가_역량기출.py
def dfs(day, money):
    global max_money
    if day > n:
        if day == n+1:
            if money > max_money:
                max_money = money
        return
    if day + a[day][0] <= n+1:
        dfs(day + a[day][0], money + a[day][1]) # 선택 o
    else:
        if money > max_money :
            max_money = money
        return
    dfs(day + 1, money) # 선택 x


n = int(input())
a = [0] + [list(map(int, input().split())) for _ in range(n)]
max_money = 0
dfs(1, 0)
print(max_money)

File: Algorithm/inflearn_cpp/myCode/84_휴가_역량기출.cpp
#include <stdio.h>
#include <vector>
using namespace std;
int n, max_money = 0;
vector<pair<int, int> > map[20];

void dfs(int day, int money)
{
    if (day == n+1)
    {
        if (money > max_money) max_money = money;
        return;
    }
    if (day + map[day][0].first <= n+1)
    {
        dfs(day + map[day][0].first, money + map[day][0].second);
    }
    dfs(day+1, money);
}

int main()
{
    int a, b;
    scanf("%d", &n);
    for (size_t i = 1; i <= n; i++)
    {
        scanf("%d %d", &a, &b);
        map[i].push_back(make_pair(a, b));
    }
    dfs(1, 0);
    printf("%d\n", max_money);
    return 0;
}